pkg/resource/clusteroperator.go

```diff
@@ -217,33 +218,44 @@ func isDeploymentStatusAvailableAndUpdated(deploy *appsapi.Deployment) bool {
 		deploy.Status.UpdatedReplicas == deploy.Status.Replicas
 }
 
-func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (modified bool, err error) {
+// syncVersions updates the version we report on ClusterOperator resource.
+//
+// We extract this version out of an Image Registry Deployment' annotation. If
+// no deployment exists(e.g. in 'Removed' state) we use the default version as
+// defined on RELEASE_VERSION environment variable.
+func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (bool, error) {
+	version := os.Getenv("RELEASE_VERSION")
+
 	deploy, err := gco.deployLister.Get(imageregistryv1.ImageRegistryName)
-	if err != nil {
-		if kerrors.IsNotFound(err) {
-			err = nil
+	if err != nil && !kerrors.IsNotFound(err) {
+		// Return error if other than "NotFound".
+		return false, err
+	} else if err == nil {
+		// Deployment exists, get the version on its annotation.
+		version = deploy.Annotations[imageregistryv1.VersionAnnotation]
+		if !isDeploymentStatusAvailableAndUpdated(deploy) {
+			return false, nil
 		}
-		return
 	}
 
-	deploymentVersion := deploy.Annotations[imageregistryv1.VersionAnnotation]
-	if len(deploymentVersion) == 0 || !isDeploymentStatusAvailableAndUpdated(deploy) {
-		return
+	if len(version) == 0 {
+		return false, fmt.Errorf("invalid empty operator version")
 	}
 
 	newVersions := []configapi.OperandVersion{
 		{
 			Name:    "operator",
-			Version: deploymentVersion,
+			Version: version,
```

Comment:
It's not explicit in the docs, but other operators can report versions for other things [1]. I propose we append the `image-registry` item to this slice if the registry has been deployed.

[1] https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md

Reply:
@adambkaplan do we want to report the "operator" version immediately then? For some reason today we report our version only when we've successfully created a deployment.

Reply:
I changed the approach to report versions(operator and image-registry) separately. PTAL.

Reply:
@dmage I seem to recall there was a lot of churn around when to report version in the 4.0 -> 4.1 timeframe.

Today I agree we should report the operator version immediately. In principle we should have `Progressing=true` during upgrade (and if not, that is a bug) [1]. Version for the image-registry shouldn't be updated until we confirm that the deployment has rolled out.

[1] https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource

Reply:
Done.